### PR TITLE
Maintenance mode for users

### PR DESCRIPTION
**Pull request by sjaanus**
1. Now maintenance mode is togglable from the UI
2. Unleash can override maintenance mode settings with a feature flag. 
3. Ordered by priority, first one taking precedence
     1. ENV_VARIABLE
     2. FEATURE_FLAG
     3. UI_SETTING

### CODE REVIEW

File: frontend/src/hooks/api/getters/useMaintenance/useMaintenance.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    const { data, error, mutate } = useSWR(
+        formatApiPath(`api/admin/maintenance`),
+        fetcher
+    );
+
+    return useMemo(
+        () => ({
+            enabled: Boolean(data?.enabled),
+            loading: !error && !data,
+            refetchMaintenance: () => mutate(),
```
Comment (kwasniew) [inline frontend/src/hooks/api/getters/useMaintenance/useMaintenance.ts:25]: just mutate

File: src/lib/app.ts [hunk 2]
```diff
@@ -139,7 +139,13 @@ export default async function getApp(
         rbacMiddleware(config, stores, services.accessService),
     );
 
-    app.use('/api/admin', maintenanceMiddleware(config));
+    app.use(
+        `${baseUriPath}/api/admin`,
+        conditionalMiddleware(
```
Comment (kwasniew) [inline src/lib/app.ts:144]: I'd move it to the controller so that we get openapi spec

File: src/lib/middleware/maintenance-middleware.ts [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
+): any => {
     const logger = getLogger('/middleware/maintenance-middleware.ts');
     logger.debug('Enabling Maintenance middleware');
 
-    return async (req: Request, res, next) => {
+    return async (req: IAuthRequest, res, next) => {
+        const maintenanceAdmin =
+            req.path.includes('/maintenance') &&
+            req.user.isAPI &&
+            req.user.permissions.includes(ADMIN);
```
Comment (kwasniew) [inline src/lib/middleware/maintenance-middleware.ts:16]: let's do it in the maintenance controller

File: src/lib/routes/admin-api/index.ts [hunk 4]
```diff
@@ -124,6 +125,14 @@ class AdminApi extends Controller {
                 new FavoritesController(config, services).router,
             ),
         );
+
+        this.app.use(
+            '/maintenance',
+            conditionalMiddleware(
```
Comment (kwasniew) [inline src/lib/routes/admin-api/index.ts:131]: can be in the controller

File: src/lib/routes/admin-api/maintenance.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    async toggleMaintenance(
+        req: IAuthRequest<unknown, unknown, MaintenanceSchema>,
+        res: Response,
+    ): Promise<void> {
+        await this.maintenanceService.toggleMaintenanceSetting(
+            req.body,
+            extractUsername(req),
+        );
+        res.status(200).end();
```
Comment (kwasniew) [inline src/lib/routes/admin-api/maintenance.ts:78]: could be 204 no content

File: src/lib/services/maintenance-service.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        settingService: SettingService,
+    ) {
+        this.config = config;
+        this.logger = config.getLogger('services/pat-service.ts');
+        this.patStore = patStore;
+        this.eventStore = eventStore;
+        this.settingService = settingService;
+    }
+
+    async getMaintenanceMode(): Promise<boolean> {
```
Comment (kwasniew) [inline src/lib/services/maintenance-service.ts:35]: isMaintenanceMode?

File: src/lib/services/maintenance-service.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    async getMaintenanceSetting(): Promise<MaintenanceSchema> {
+        return (
+            (await this.settingService.get(maintenanceSettingsKey)) || {
+                enabled: false,
+            }
+        );
+    }
+
+    async toggleMaintenanceSetting(
```
Comment (kwasniew) [inline src/lib/services/maintenance-service.ts:50]: toggleMaintenanceMode

File: src/test/e2e/api/admin/maintenance.e2e.test.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+afterEach(async () => {
+    await db.stores.featureToggleStore.deleteAll();
+});
+
+afterAll(async () => {
+    await db.destroy();
+});
+
+test('should have access to the get all features endpoint even if api is disabled', async () => {
```
Comment (kwasniew) [inline src/test/e2e/api/admin/maintenance.e2e.test.ts:19]: can be deleted

File: src/test/e2e/api/admin/maintenance.e2e.test.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+            disableLegacyFeaturesApi: true,
+        },
+    );
+
+    await appWithDisabledLegacyFeatures.request
+        .get('/api/admin/features')
+        .expect(200);
+});
+
+test('should not allow creation of feature toggle in maintenance mode', async () => {
```
Comment (kwasniew) [inline src/test/e2e/api/admin/maintenance.e2e.test.ts:32]: should not allow to create feature toggles in ...

File: src/test/e2e/api/admin/maintenance.e2e.test.ts [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    return appWithMaintenanceMode.request
+        .post('/api/admin/features')
+        .send({
+            name: 'maintenance-feature',
+        })
+        .set('Content-Type', 'application/json')
+        .expect(503);
+});
+
+test('should not go into maintenance, when feature is fully turned off', async () => {
```
Comment (kwasniew) [inline src/test/e2e/api/admin/maintenance.e2e.test.ts:51]: should not go into maintenance mode when maintenance feature is off

File: src/test/e2e/api/admin/maintenance.e2e.test.ts [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    return appWithMaintenanceMode.request
+        .post('/api/admin/features')
+        .send({
+            name: 'maintenance-feature1',
+        })
+        .set('Content-Type', 'application/json')
+        .expect(201);
+});
+
+test('should not go into maintenance, when user has not set it', async () => {
```
Comment (kwasniew) [inline src/test/e2e/api/admin/maintenance.e2e.test.ts:70]: maintenance mode is off by default

File: src/test/e2e/api/admin/maintenance.e2e.test.ts [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    const appWithMaintenanceMode = await setupAppWithCustomConfig(db.stores, {
+        experimental: {
+            flags: {
+                maintenance: true,
+                maintenanceMode: false,
```
Comment (kwasniew) [inline src/test/e2e/api/admin/maintenance.e2e.test.ts:75]: let's remove this line

File: src/test/e2e/api/admin/maintenance.e2e.test.ts [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    return appWithMaintenanceMode.request
+        .post('/api/admin/features')
+        .send({
+            name: 'maintenance-feature1',
+        })
+        .set('Content-Type', 'application/json')
+        .expect(201);
+});
+
+test('should go into maintenance, when user has set it', async () => {
```
Comment (kwasniew) [inline src/test/e2e/api/admin/maintenance.e2e.test.ts:89]: into maintenance mode

File: src/test/e2e/api/admin/maintenance.e2e.test.ts [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    await appWithMaintenanceMode.request
+        .post('/api/admin/maintenance')
+        .send({
+            enabled: true,
+        })
+        .set('Content-Type', 'application/json')
+        .expect(404);
+});
+
+test('feature flag should take precedence over setting', async () => {
```
Comment (kwasniew) [inline src/test/e2e/api/admin/maintenance.e2e.test.ts:134]: maintenance mode flag should take precedence over maintenance mode setting

**Review by kwasniew — APPROVED** [review]

File: src/lib/app.ts (hunk 2 above)
Comment (sjaanus) [reply src/lib/app.ts:144]: This is just middleware for blocking requests, not adding any route.

**Pull request merged**